### PR TITLE
IAM support for AARC-G002 profile

### PR DESCRIPTION
**Pull request by enricovianello**

### CODE REVIEW

File: iam-login-service/src/main/java/it/infn/mw/iam/config/IamProperties.java [hunk 1]
```diff
@@ -230,6 +231,20 @@ public void setLocation(String location) {
     }
   }
 
+  public static class AarcProfileProperties {
+
+    private String urnNamespace = "geant:iam";
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/config/IamProperties.java]: I notice this now. Why geant here?
I would have infn as a default value.

Comment (andreaceccanti) [reply iam-login-service/src/main/java/it/infn/mw/iam/config/IamProperties.java]: As suggested probably using 'example' is better.

Comment (enricovianello) [reply iam-login-service/src/main/java/it/infn/mw/iam/config/IamProperties.java]: waiting for an 'infn' entry, example is listed here:
https://www.iana.org/assignments/urn-namespaces/urn-namespaces.xhtml
then 'example' is a good temporary solution. Fixed.

File: iam-login-service/src/test/java/it/infn/mw/iam/test/oauth/profile/AarcClaimValueHelperTests.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    IamGroup parent = buildGroup("parent");
+    IamGroup child = buildGroup("child", parent);
+    when(userInfo.getGroups()).thenReturn(Sets.newHashSet(parent, child));
+
+    Set<String> urns = helper.resolveGroups(userInfo);
+    assertThat(urns, hasSize(2));
+    assertThat(urns, hasItem(parentUrn));
+    assertThat(urns, hasItem(childUrn));
+  }
+
```
[the remark this answers is not in the document]
Comment (enricovianello) [reply iam-login-service/src/test/java/it/infn/mw/iam/test/oauth/profile/AarcClaimValueHelperTests.java]: The required test with no groups has been added.

File: iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/profile/aarc/AarcClaimValueHelper.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    StringBuilder groupHierarchy = new StringBuilder(group.getName());
+    Optional<IamGroup> parent = Optional.ofNullable(group.getParentGroup());
+    while (parent.isPresent()) {
+      groupHierarchy.insert(0, parent.get().getName() + ":");
+      parent = Optional.ofNullable(parent.get().getParentGroup());
+    }
+    urn.append(groupHierarchy.toString());
+
+    urn.append("#");
+    urn.append(organisationName);
```
[the remark this answers is not in the document]
Comment (enricovianello) [reply iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/profile/aarc/AarcClaimValueHelper.java]: Fixed. My mistake in understanding. I can't find a IAM property for the https port. At the moment I hard-coded '443' to the 'iam.host' property.

File: iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/profile/aarc/AarcClaimValueHelper.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      .getGroups()
+      .forEach(g -> encodedGroups.add(encodeGroup(g)));
+    return encodedGroups;
+  }
+
+  private String encodeGroup(IamGroup group) {
+
+    StringBuilder urn = new StringBuilder();
+
+    urn.append("urn:");
```
[the remark this answers is not in the document]
Comment (enricovianello) [reply iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/profile/aarc/AarcClaimValueHelper.java]: Done.

File: iam-login-service/src/main/java/it/infn/mw/iam/config/IamProperties.java [hunk 5]
```diff
@@ -264,6 +279,8 @@ public void setLocation(String location) {
 
   private boolean generateDdlSqlScript = false;
 
+  private Urn urn = new Urn();
```
[the remark this answers is not in the document]
Comment (enricovianello) [reply iam-login-service/src/main/java/it/infn/mw/iam/config/IamProperties.java]: Done.

File: iam-login-service/src/main/java/it/infn/mw/iam/config/IamProperties.java [hunk 6]
```diff
@@ -230,6 +231,20 @@ public void setLocation(String location) {
     }
   }
 
+  public static class Urn {
+
+    private String namespace = "geant:iam";
```
[the remark this answers is not in the document]
Comment (enricovianello) [reply iam-login-service/src/main/java/it/infn/mw/iam/config/IamProperties.java]: Done.

File: iam-login-service/src/main/java/it/infn/mw/iam/config/IamProperties.java [hunk 7]
```diff
@@ -230,6 +231,20 @@ public void setLocation(String location) {
     }
   }
 
+  public static class Urn {
```
[the remark this answers is not in the document]
Comment (enricovianello) [reply iam-login-service/src/main/java/it/infn/mw/iam/config/IamProperties.java]: Done.

File: iam-persistence/src/main/resources/db/migration/test/V100000_6__add_aarc_test_client_and_scopes.sql [hunk 8]
```diff
@@ -0,0 +1,7 @@
+INSERT INTO system_scope(scope, description, icon, restricted, default_scope, structured, structured_param_description) VALUES
+  ('eduperson_scoped_affiliation', 'aarc organisation name', 'null', false, false, false, null),
+  ('eduperson_entitlement', 'aarc groups', 'null', false, false, false, null);
```
Comment (andreaceccanti) [inline iam-persistence/src/main/resources/db/migration/test/V100000_6__add_aarc_test_client_and_scopes.sql]: Change descriptions to:

- eduperson_scoped_affiliation:  "Access to EduPerson scoped affiliation information"
- eduperson_entitlement: "Access to EduPerson entitlements information"

Comment (enricovianello) [reply iam-persistence/src/main/resources/db/migration/test/V100000_6__add_aarc_test_client_and_scopes.sql]: Done. And also moved this INSERT into both h2 and mysql migrations.

**Pull request merged**